packages/react/context-menu/src/ContextMenu.tsx

```diff
@@ -18,17 +25,53 @@ const CONTEXT_MENU_NAME = 'ContextMenu';
 type ContextMenuContextValue = {
   open: boolean;
   onOpenChange(open: boolean): void;
+  onRootContentRef(node: HTMLDivElement): void;
 };
 
 const [ContextMenuProvider, useContextMenuContext] = createContext<ContextMenuContextValue>(
   CONTEXT_MENU_NAME
 );
 
-const ContextMenu: React.FC<{ onOpenChange?(open: boolean): void }> = (props) => {
+type ContextMenuCommonOwnProps = {
+  open?: boolean;
+  defaultOpen?: boolean;
+  onOpenChange?(open: boolean): void;
+};
+
+const ContextMenu: React.FC<
+  ContextMenuCommonOwnProps & {
+    dir?: Direction;
+  }
+> = (props) => {
+  const { dir, ...menuProps } = props;
+  const isInsideParentMenu = React.useContext(NestedMenuContext);
+  const isInsideParentContent = React.useContext(InsideContentContext);
+  const parentDirection = React.useContext(DirectionContext);
+
+  return (
+    <NestedMenuContext.Provider
+      value={isInsideParentMenu === undefined || !isInsideParentContent ? false : true}
+    >
+      <DirectionContext.Provider value={parentDirection || dir}>
+        <ContextMenuImpl {...menuProps} />
+      </DirectionContext.Provider>
+    </NestedMenuContext.Provider>
+  );
+};
+
+const ContextMenuImpl: React.FC<ContextMenuCommonOwnProps> = (props) => {
+  const isSubmenu = React.useContext(NestedMenuContext);
+  return isSubmenu ? <MenuPrimitive.Sub {...props} /> : <ContextMenuRoot {...props} />;
+};
+
+const ContextMenuRoot: React.FC<ContextMenuCommonOwnProps> = (props) => {
   const { children, onOpenChange } = props;
   const [open, setOpen] = React.useState(false);
-  const handleOpenChangeProp = useCallbackRef(onOpenChange);
+  const [rootContentElement, setRootContentElement] = React.useState<HTMLDivElement | null>(null);
+  const directionContext = React.useContext(DirectionContext);
+  const direction = useDirection(rootContentElement, directionContext);
```

Comment:
Performing the style query for `direction` at the root level only and threading it down via context. I couldn't find a good enough reason to take a perf hit and do it JIT for every sub menu.